README
python-purple - Python bindings for libpurple
---------------------------------------------

Copyright (c) 2008 INdT - Instituto Nokia de Tecnologia

Requirements
------------

In order to build python-purple, you need the following packages:
 * cdbs
 * debhelper
 * libpurple-dev
 * libglib2.0-dev
 * python2.5-dev
 * python2.5-distutils
 * cython (optional)

Although it's not dependent, cython should also needed if you want to
re-create purple.c file (generated from cython using purple.pyx).

Building and installing python-purple
-------------------------------------

Building cython-generated purple.c file (optional):

$ cython purple.pyx -I libpurple/

This is also done by setup.py when running:

$ python2.5 setup.py build

If you don't need to generate purple.c file, you can use setup_dist.py:

$ python2.5 setup_dist.py build

Both setups generates purple.so inside build/ directory. The difference is
that setup.py also generates purple.c file, if it doesn't exists.

To install files inside your system:

$ sudo python2.5 setup.py install --root=/usr

You can also create a debian package of it:

$ dpkg-buildpackage -rfakeroot

Testing python-purple
---------------------

Python-purple provides a client example (nullclient) which connects to a
Gtalk account and displays libpurple's debug messages. In order to execute it,
issue the following command:

$ LD_PRELOAD=/usr/lib/libpurple.so.0 python2.5 nullclient.py

It will ask for a username and then password. Insert it and it will
automatically connect to the given account.

There is also another example client (nullclient-ecore), which provides some
callback examples from libpurple interacting with UI. In order to execute it,
issue the following command:

$ LD_PRELOAD=/usr/lib/libpurple.so.0 python2.5 nullclient-ecore.py

An Etk-based window will pop up showing text windows, a command entry and
three buttons ('Send', 'Connect', 'Add account'). In order to test it, follow
the steps below:

1. Click on 'Add account' button. A pop-up window appears, and you can enter
there a gmail-based account (e.g. jsmith@gmail.com). Click 'Ok' when done.

2. Click on 'Connect' button. A pop-up window appears asking for account's
password. Click 'Ok' when done.

Now wait for the account connection. When 'Connected' appears, your buddy
list gets updated.

3. You are now connected! Select a buddy from buddy list, write a message on
the command entry and click 'Send'. Your message appears in the text box.
If your buddy replies, his/her message also gets shown in the text box.


debian/changelog
python-purple (0.1ubuntu1) jaunty; urgency=low

  * Removed python2.5-distutils dependency for Ubuntu Jaunty.
  * Removed LD_PRELOAD dependency.
  * Replaced Ecore mainloop with GLib.

 -- Bruno Abinader <bruno.abinader@openbossa.org>  Tue, 16 Jun 2009 10:16:37 -0400

python-purple (0.1) unstable; urgency=low

  * Initial release.

 -- Bruno Abinader <bruno.abinader@openbossa.org>  Tue, 11 Nov 2008 16:53:59 -0400


c_purple.c
/*
 * Copyright (c) 2008 INdT - Instituto Nokia de Tecnologia
 *
 * This file is part of python-purple.
 *
 * python-purple is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * python-purple is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <purple.h>

#include <glib.h>

#include <signal.h>
#include <string.h>
#include <unistd.h>

#include "c_purple.h"

#define PURPLE_GLIB_READ_COND  (G_IO_IN | G_IO_HUP | G_IO_ERR)
#define PURPLE_GLIB_WRITE_COND (G_IO_OUT | G_IO_HUP | G_IO_ERR | G_IO_NVAL)

typedef struct _PurpleGLibIOClosure {
	PurpleInputFunction function;
	guint result;
	gpointer data;
} PurpleGLibIOClosure;

static gboolean purple_glib_io_invoke(GIOChannel *source, GIOCondition condition,
		gpointer data)
{
	PurpleGLibIOClosure *closure = data;
	PurpleInputCondition purple_cond = 0;

	if (condition & PURPLE_GLIB_READ_COND)
		purple_cond |= PURPLE_INPUT_READ;
	if (condition & PURPLE_GLIB_WRITE_COND)
		purple_cond |= PURPLE_INPUT_WRITE;

	closure->function(closure->data, g_io_channel_unix_get_fd(source),
			purple_cond);

	return TRUE;
}

guint glib_input_add(gint fd, PurpleInputCondition condition, PurpleInputFunction function,
		gpointer data)
{
	PurpleGLibIOClosure *closure = g_new0(PurpleGLibIOClosure, 1);
	GIOChannel *channel;
	GIOCondition cond = 0;

	closure->function = function;
	closure->data = data;

	if (condition & PURPLE_INPUT_READ)
		cond |= PURPLE_GLIB_READ_COND;
	if (condition & PURPLE_INPUT_WRITE)
		cond |= PURPLE_GLIB_WRITE_COND;

	channel = g_io_channel_unix_new(fd);
	closure->result = g_io_add_watch_full(channel, 0, cond,
			purple_glib_io_invoke, closure, g_free);

	g_io_channel_unref(channel);
	return closure->result;
}

static void
signed_on(PurpleConnection *gc, gpointer null)
{
	PurpleAccount *account = purple_connection_get_account(gc);
	printf("Account connected: %s %s\n", account->username, account->protocol_id);
}

void connect_to_signals_for_demonstration_purposes_only(void)
{
	static int handle;

	purple_signal_connect(purple_connections_get_handle(), "signed-on", &handle,
				PURPLE_CALLBACK(signed_on), NULL);
}


c_purple.h
/*
 * Copyright (c) 2008 INdT - Instituto Nokia de Tecnologia
 *
 * This file is part of python-purple.
 *
 * python-purple is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * python-purple is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

guint glib_input_add(gint fd, PurpleInputCondition condition,
	PurpleInputFunction function, gpointer data);

void connect_to_signals_for_demonstration_purposes_only(void);


nullclient.py
#
#  Copyright (c) 2008 INdT - Instituto Nokia de Tecnologia
#
#  This file is part of python-purple.
#
#  python-purple is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  python-purple is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import getpass
import sys, dl
import time

sys.setdlopenflags(dl.RTLD_NOW | dl.RTLD_GLOBAL)

import purple

# The information below is needed by libpurple
__NAME__ = "nullclient"
__VERSION__ = "0.1"
__WEBSITE__ = "N/A"
__DEV_WEBSITE__ = "N/A"

if __name__ == '__main__':
    # Sets initial parameters
    core = purple.Purple(__NAME__, __VERSION__, __WEBSITE__, __DEV_WEBSITE__, \
            debug_enabled=True, default_path="/tmp")

    # Initializes libpurple
    core.purple_init()

    # Get username from user
    sys.stdout.write("Enter GTalk account: ")
    username = sys.stdin.readline()[:-1]

    # Initialize protocol class
    protocol = purple.Protocol('prpl-jabber')

    # Creates new account inside libpurple
    account = purple.Account(username, protocol, core)
    account.new()

    # Get password from user
    account.set_password(getpass.getpass())

    # Set account protocol options
    info = {}
    info['connect_server'] = 'talk.google.com'
    info['port'] = '443'
    info['old_ssl'] = True
    account.set_protocol_options(info)

    # Enable account (connects automatically)
    account.set_enabled(True)

    while True:
        try:
            core.iterate_main_loop()
            time.sleep(0.01)
        except KeyboardInterrupt:
            core.destroy()
            break


setup_dist.py
#!/usr/bin/env python

from distutils.core import setup
from distutils.extension import Extension

from subprocess import Popen, PIPE

cflags = Popen(['pkg-config', '--cflags', 'purple'], stdout=PIPE).communicate()[0].split()
ldflags = Popen(['pkg-config', '--libs', 'purple'], stdout=PIPE).communicate()[0].split()

purplemodule = Extension('purple',
                         sources=['c_purple.c','purple.c'],
                         extra_compile_args=cflags,
                         extra_link_args=ldflags)

long_description = "\
Python bindings for libpurple, a multi-protocol instant messaging library."

setup(name = 'python-purple',
      version = '0.1',
      author ='Bruno Abinader',
      author_email = 'bruno.abinader@openbossa.org',
      description = 'Python bindings for Purple',
      long_description = long_description,
      ext_modules = [purplemodule],
      )


nullclient-ecore.py
#
#  Copyright (c) 2008 INdT - Instituto Nokia de Tecnologia
#
#  This file is part of python-purple.
#
#  python-purple is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  python-purple is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import etk
import ecore
import sys, dl

sys.setdlopenflags(dl.RTLD_NOW | dl.RTLD_GLOBAL)

import purple

# The information below is needed by libpurple
__NAME__ = "nullclient-ecore"
__VERSION__ = "0.1"
__WEBSITE__ = "N/A"
__DEV_WEBSITE__ = "N/A"

class MainWindow(object):
    def __init__(self):
        # Connect button callback
        self.__connect_button_cb = None

        # Send button callback
        self.__send_button_cb = None

        # New account callback
        self.__new_account_cb = None

        # Exit callback
        self.__exit_cb = None

    def __login_window_cb(self, pointer):
        # Password entry
        self.login_password = etk.Entry()

        # Confirm login button
        confirm_login_bt = etk.Button(label='Ok')
        confirm_login_bt.on_clicked(self.__connect_button_clicked)

        # Login VBox
        vbox = etk.VBox()
        vbox.append(self.login_password, etk.VBox.START, etk.VBox.FILL, 0)
        vbox.append(confirm_login_bt, etk.VBox.END, etk.VBox.NONE, 0)

        # Login window
        self.login_window = etk.Window(title='Password', \
                size_request=(190, 80), child=vbox)
        self.login_window.show_all()

    def __connect_button_clicked(self, pointer):
        # Call connection button callback
        if self.__connect_button_cb:
            self.__connect_button_cb(self.login_password.text)
        self.login_window.destroy()

    def __send_button_clicked(self, pointer):
        # Get selected buddy name from buddy list
        bname = self.blist.selected_rows[0][0]

        # Get message from command entry
        msg = self.cmd_entry.text

        if bname and msg != '':
            # Call send button callback
            if self.__send_button_cb:
                self.__send_button_cb(bname, msg)

        # Clear text from command entry
        self.cmd_entry.text = ''

    def __new_account_button_clicked(self, pointer):
        # Username entry
        self.login_username = etk.Entry()

        # Confirm username button
        confirm_username_bt = etk.Button(label='Ok')
        confirm_username_bt.on_clicked(self.add_account)

        # Username VBox
        vbox = etk.VBox()
        vbox.append(self.login_username, etk.VBox.START, etk.VBox.FILL, 0)
        vbox.append(confirm_username_bt, etk.VBox.END, etk.VBox.NONE, 0)

        # Username window
        self.username_window = etk.Window(title='Username', \
                size_request=(190, 80), child=vbox)
        self.username_window.show_all()

    def __create_accounts_list(self):
        # Accounts list
        self.accslistmodel = etk.ListModel()
        self.accslist = etk.List(model=self.accslistmodel, \
                columns=[(10, etk.TextRenderer(slot=0), False)], \
                selectable=True, animated_changes=True)

        #Appending accounts list to VBox
        vbox = etk.VBox()
        vbox.append(self.accslist, etk.VBox.START, etk.VBox.EXPAND_FILL, 0)
        return vbox

    def __create_buddies_list(self):
        # Buddies list
        self.blistmodel = etk.ListModel()
        self.blist = etk.List(model=self.blistmodel, \
                columns=[(10, etk.TextRenderer(slot=0), False)], \
                selectable=True, animated_changes=True)

        # Appending buddies list to VBox
        vbox = etk.VBox()
        vbox.append(self.blist, etk.VBox.START, etk.VBox.EXPAND_FILL, 0)
        return vbox

    def __create_buttons_bar(self):
        # Send button
        send_button = etk.Button(label='Send')
        send_button.on_clicked(self.__send_button_clicked)

        # Connect button
        conn_button = etk.Button(label='Connect')
        conn_button.on_clicked(self.__login_window_cb)

        # New account button
        new_acc_button = etk.Button(label='New Account')
        new_acc_button.on_clicked(self.__new_account_button_clicked)

        # Appending all buttons to HBox
        hbox = etk.HBox(homogeneous=False)
        hbox.append(send_button, etk.HBox.START, etk.HBox.NONE, 0)
        hbox.append(conn_button, etk.HBox.START, etk.HBox.NONE, 0)
        hbox.append(new_acc_button, etk.HBox.START, etk.HBox.NONE, 0)
        return hbox

    def __create_command_entry_box(self):
        # Command entry box
        self.cmd_entry = etk.Entry()
        self.cmd_label = etk.Label(text='Type your message: ')

        # appending command entry and label to HBox
        hbox = etk.HBox(homogeneous=False)
        hbox.append(self.cmd_label, etk.HBox.START, \
                etk.HBox.START, 0)
        hbox.append(self.cmd_entry, etk.HBox.START, \
                etk.HBox.EXPAND_FILL, 0)
        return hbox

    def __create_text_area(self):
        # Text area (shows buddy messages)
        self.txt_area = etk.Label()
        self.txt_area.text = '<br>Nullclient-Ecore<br> '

        # Appending text area to VBox
        vbox = etk.VBox()
        vbox.append(self.txt_area, etk.VBox.START, etk.VBox.EXPAND_FILL, 0)
        return vbox

    def __create_main_panel(self):
        # Text box
        txt_vbox = self.__create_text_area()

        # Buddies list
        bdd_vbox = self.__create_buddies_list()

        # Accounts list
        acc_vbox = self.__create_accounts_list()

        # Appending text area, buddies list and accounts list to HBox
        hbox = etk.HBox()
        hbox.append(txt_vbox, etk.HBox.START, etk.HBox.EXPAND_FILL, 0)
        hbox.append(bdd_vbox, etk.HBox.END, etk.HBox.EXPAND_FILL, 0)
        hbox.append(acc_vbox, etk.HBox.END, etk.HBox.EXPAND_FILL, 0)
        return hbox

    def __create_main_box(self):
        # Main panel
        panel_hbox = self.__create_main_panel()

        # Command entry
        cmd_hbox = self.__create_command_entry_box()

        # Buttons Bar
        btn_hbox = self.__create_buttons_bar()

        # Connection status
        self.status = etk.Label(text='Connection status')

        # Main VBox
        vbox = etk.VBox(homogeneous=False)
        vbox.append(panel_hbox, etk.VBox.START, etk.VBox.EXPAND_FILL, 0)
        vbox.append(cmd_hbox, etk.VBox.END, etk.VBox.FILL, 0)
        vbox.append(btn_hbox, etk.VBox.END, etk.VBox.NONE, 5)
        vbox.append(self.status, etk.VBox.END, etk.VBox.FILL, 0)
        return vbox

    def get_selected_account(self):
        # Catch selected account from accounts list
        try:
            account = self.accslist.selected_rows[0][0]
            if account:
                return account
            else:
                return None
        except:
            return None

    def add_buddy(self, name):
        # Adds a new buddy into buddy list
        if [name] not in self.blistmodel.elements:
            self.blistmodel.append([name])

    def remove_buddy(self, name):
        # Removes a buddy from buddy list
        self.blistmodel.remove([name])

    def add_account(self, pointer):
        # Adds a new account into accounts list
        if [self.login_username.text] not in self.accslistmodel.elements:
            self.accslistmodel.append([self.login_username.text])
        self.username_window.destroy()
        self.window.show_all()

    def add_connection_button_cb(self, cb):
        if callable(cb):
            self.__connect_button_cb = cb

    def add_new_account_cb(self, cb):
        if callable(cb):
            self.__new_account_cb = cb

    def add_send_button_cb(self, cb):
        if callable(cb):
            self.__send_button_cb = cb

    def add_exit_cb(self, cb):
        if callable(cb):
            self.__exit_cb = cb

    def init_window(self):
        # Main box
        main_box = self.__create_main_box()

        # Main Window
        self.window = etk.Window(title='Nullclient-Ecore', \
                size_request=(600, 600), child=main_box)
        self.window.on_destroyed(self.__exit_cb)
        self.window.show_all()

    def show(self):
        if self.window:
            self.window.show_all()

class NullClient(object):
    def __init__(self):
        # Sets initial parameters
        self.core = purple.Purple(__NAME__, __VERSION__, __WEBSITE__, \
                __DEV_WEBSITE__, debug_enabled=True, default_path='/tmp')
        ecore.Timer(0.01, self.core.iterate_main_loop)
        self.account = None
        self.buddies = {}
        self.conversations = {}
        self.protocol = purple.Protocol('prpl-jabber')
        self.window = MainWindow()

        # Adds libpurple core callbacks

        # Updates buddy list
        self.core.add_callback('blist', 'update', \
                self.__update_blist_cb)

        # Updates connection progress
        self.core.add_callback('connection', 'connect-progress', \
                self.__connection_progress_cb)

        # Activates when an account is connected
        self.core.add_callback('connection', 'connected', \
                self.__connected_cb)

        # Activates when an account is disconnected
        self.core.add_callback('connection', 'disconnected', \
                self.__disconected_cb)

        # Activates when a message is sent or received from conversation
        self.core.add_callback('conversation', 'write-im', \
                self.__write_im_cb)

        # Signal when account signed on
        self.core.signal_connect('signed-on', self.__signed_on_cb)

        # Signal when account signed off
        self.core.signal_connect('signed-off', self.__signed_off_cb)

        # Signal when buddy signed on
        self.core.signal_connect('buddy-signed-on', self.__buddy_signed_on_cb)

        # Signed when buddy signed off
        self.core.signal_connect('buddy-signed-off', self.__buddy_signed_off_cb)

        # Adds UI callbacks
        self.window.add_connection_button_cb(self.connect)
        self.window.add_send_button_cb(self.send_message)
        self.window.add_exit_cb(self.exit)

        # Initializes libpurple
        self.core.purple_init()

        # Initializes UI
        self.window.init_window()

    def __update_blist_cb(self, type, name=None, alias=None):
        if self.account and name and type == 2:
            if name not in self.buddies:
                self.buddies[name] = purple.Buddy(name, self.account)
            if self.buddies[name].online:
                self.window.add_buddy(name)

    def __connection_progress_cb(self, text, step, step_count):
        if self.window:
            self.window.status.text = text

    def __connected_cb(self, *data):
        if self.window:
            self.window.status.text = 'Connected'

    def __disconected_cb():
        if self.window:
            self.window.status.text = 'Disconnected'

    def __write_im_cb(self, username, name, alias, message, flags):
        if self.window:
            if 'SEND' == flags:
                self.window.txt_area.text += username + ": " + message + "<br> "
            elif alias:
                self.window.txt_area.text += alias + ": " + message + "<br> "
            else:
                self.window.txt_area.text += name + ": " + message + "<br> "
            self.window.show()

    def __signed_on_cb(self, username, protocol_id):
        if self.window:
            self.window.txt_area += 'Signed on: %s (%s)' % (username, protocol_id)
            self.window.show()

    def __signed_off_cb(self, username, protocol_id):
        if self.window:
            self.window.txt_area += 'Signed off: %s (%s)' % (username, protocol_id)
            self.window.show()

    def __buddy_signed_on_cb(self, name, alias):
        if self.window:
            self.window.txt_area += 'Buddy signed on: %s (%s)' % (name, alias)
            self.window.show()

    def __buddy_signed_off_cb(self, name, alias):
        if name in self.buddies:
            del self.buddies[name]

        if self.window:
            self.window.txt_area += 'Buddy signed off: %s (%s)' % (name, alias)
            self.window.remove_buddy(name)
            self.window.show()

    def connect(self, password):
        username = self.window.get_selected_account()
        if username and password:
            self.account = purple.Account(username, self.protocol, self.core)
            if not self.account.exists:
                self.account.new()
                info = {}
                info['connect_server'] = 'talk.google.com'
                info['port'] = '443'
                info['old_ssl'] = True
                self.account.set_protocol_options(info)

            self.account.set_password(password)
            self.account.set_enabled(True)

    def send_message(self, name, message):
        print name, message
        if name not in self.conversations:
            self.conversations[name] = purple.Conversation('IM', self.account, name)
            self.conversations[name].new()

        self.conversations[name].im_send(message)

    def exit(self, pointer):
        ecore.main_loop_quit()

if __name__ == '__main__':
    client = NullClient()

    # Initializes ecore mainloop
    ecore.main_loop_begin()
